numba/np/arrayobj.py

```diff
@@ -1382,6 +1382,64 @@ def impl(array, shape):
     return impl
 
 
+@register_jitable
+def numpy_broadcast_shapes_list(r, m, shape):
+    for i in range(len(shape)):
+        k = m - len(shape) + i
+        tmp = shape[i]
+        if tmp == 1:
+            continue
+        if r[k] == 1:
+            r[k] = tmp
+        elif r[k] != tmp:
+            raise ValueError("shape mismatch: objects"
+                             " cannot be broadcast"
+                             " to a single shape")
+
+
+@overload(np.broadcast_arrays)
+def numpy_broadcast_arrays(*args):
+
+    for idx, arg in enumerate(args):
+        if not type_can_asarray(arg):
+            raise errors.TypingError(f'Argument "{idx}" must '
+                                     'be array-like')
+
+    # number of dimensions
+    m = 0
+    for idx, arg in enumerate(args):
+        if isinstance(arg, types.ArrayCompatible):
+            m = max(m, arg.ndim)
+        elif isinstance(arg, (types.Number, types.StringLiteral,
+                              types.Boolean)):
```

Comment:
Bool inputs are not tested? This is failing for me:
```python
from numba import njit
import numpy as np

@njit
def foo(x,):
    return np.broadcast_arrays(x, True)

b = np.arange(6).reshape((2, 3))
print(foo(b,))
```
the internal list is typed for int array but a bool array is being appended. This raises the question... does this implementation only work if the inputs are homogeneous in dtype? I think this might be the case, e.g.
```python
from numba import njit
import numpy as np

@njit
def foo(x, y):
    return np.broadcast_arrays(x, y)

a = np.arange(3, dtype=np.complex128)
b = np.arange(6).reshape((2, 3))
print(foo(a, b))
```
to which effect this probably ought to be caught at typing time or fixed in the implementation?

Reply:
The code above crashes because the current implementation appends each intermediary result to a list. Is there any way one could append different types to the list (or tuple) without unifying them?

Reply:
I think invariant of the implementation, the issue is that the return type of the NumPy function is a list of lists with no guarantee that the nested list(s) are of the same type? Numba cannot construct such a list dynamically, the list must be homogeneous in type. Perhaps we could discuss this at the developer's meeting or similar, I don't think there's a work around at present.

Reply:
Could this be done using another data structure? I know one can return a heterogeneous tuple from a jitted function.

Reply:
Technically I think it could be done, but it would break the semantics of the NumPy function as the return type would not match that of NumPy. I'm not sure that there's any reason for this to be a list other than the list comprehension that generates it is idiomatic:
https://github.com/numpy/numpy/blob/73e48ae177c4b29cdbdbcfd39e22cf5de6c7f125/numpy/lib/stride_tricks.py#L546-L547

Reply:
Got it! If breaking the API is not a problem, I think I can do it with tuples.

Reply:
Unfortunately, for Numba, breaking the NumPy API is a problem. Perhaps raise an issue/make a suggestion on NumPy's issue tracker about it returning a tuple?

Reply:
I don't think this will change in the near future. The array-api spec also says that `broadcast_arrays` return type is a `list[array]`:
https://data-apis.org/array-api/latest/API_specification/generated/signatures.data_type_functions.broadcast_arrays.html?highlight=broadcast_arrays

Reply:
RE data-api spec, I suppose the question is: was this explicitly designed or was it inherited from/informed by NumPy semantics? Seems like the result is a fixed size that is derived solely from number of args so maybe a tuple would be a better fit? It certainly helps compilers with respect to typing/type stability if fixed sized containers are used where possible.

For this PR. I agree, this can be implemented with the homogeneous type restriction.

Reply:
> For this PR. I agree, this can be implemented with the homogeneous type restriction.

Do you have any more comments about the PR?

Reply:
Once this is fixed I'll retest it, main thing is to ensure that only types/combinations of types that are supported are permitted through and that the result is correct in value and type! More testing/checking in this area is welcomed. Thanks.

Reply:
The issue with the example you posted is because you're mixing different types. The following works

```python
from numba import njit
import numpy as np

@njit
def foo(x,):
    return np.broadcast_arrays(x, False)

b = np.full(6, fill_value=True).reshape((2, 3))
print(foo(b,))
```

Reply:
Apologies, I wasn't very clear. Fixing this means ensuring that type guards are present to prevent the code running with the unsupported "mixing different types".